propagate third party claims for id tokens

jimmy1wu opened the pull request:
for https://github.com/OpenLiberty/open-liberty/issues/16298

jimmy1wu committed:
propagate 3rd party id token claims

jimmy1wu commented:
#libby

jimmy1wu commented:
#build
#spawn.fullfat.buckets=com.ibm.ws.security.oidc.client_fat.1,com.ibm.ws.security.oidc.client_fat.2,com.ibm.ws.security.oidc.client_fat.config,com.ibm.ws.security.oidc.client_fat.e2e,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.saml,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat,com.ibm.ws.security.oidc.server_fat.app-passwords,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.clientregistration,com.ibm.ws.security.oidc.server_fat.config,com.ibm.ws.security.oidc.server_fat.config.oauth,com.ibm.ws.security.oidc.server_fat.config.oidc,com.ibm.ws.security.oidc.server_fat.endpoint.authorize,com.ibm.ws.security.oidc.server_fat.endpoint.coveragemap,com.ibm.ws.security.oidc.server_fat.endpoint.discovery,com.ibm.ws.security.oidc.server_fat.endpoint.end_session,com.ibm.ws.security.oidc.server_fat.endpoint.introspect,com.ibm.ws.security.oidc.server_fat.endpoint.revoke,com.ibm.ws.security.oidc.server_fat.endpoint.token,com.ibm.ws.security.oidc.server_fat.endpoint.userinfo,com.ibm.ws.security.oidc.server_fat.feature,com.ibm.ws.security.oidc.server_fat.granttype.jwt,com.ibm.ws.security.oidc.server_fat.gui.tokenMgmt,com.ibm.ws.security.oidc.server_fat.jaxrs,com.ibm.ws.security.oidc.server_fat.jaxrs.config,com.ibm.ws.security.oidc.server_fat.oauth,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.oidc.server_fat.saml,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP.1,com.ibm.ws.security.social_fat.LibertyOP.2,com.ibm.ws.security.social_fat.OpenShift,com.ibm.ws.security.social_fat.delegated,com.ibm.ws.security.social_fat.multiProvider

jantley-ibm commented on dev/com.ibm.ws.security.openidconnect.server/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -168,6 +168,12 @@ requireRequestUriRegistration.desc=Indicate by true or false whether require req
 customClaims=Extra claims of ID token
 customClaims.desc=The extra claims to be put in the payloads of the ID Token, in addition to the default realmName, uniqueSecurityName and groupIds claims.
 
+thirdPartyIDTokenClaims=Extra claims from a third-party ID Token
```
change
ID Token
to
ID token

(suggest also making same change on line 169)

jantley-ibm commented on dev/com.ibm.ws.security.openidconnect.server/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -168,6 +168,12 @@ requireRequestUriRegistration.desc=Indicate by true or false whether require req
 customClaims.desc=The extra claims to be put in the payloads of the ID Token, in addition to the default realmName, uniqueSecurityName and groupIds claims.
 
+thirdPartyIDTokenClaims=Extra claims from a third-party ID Token
+thirdPartyIDTokenClaims.desc=The extra claims from a third-party ID Token to be put in the payload of the ID Token.
```
change
ID Token
to
ID token
(2 places)

jantley-ibm commented on dev/com.ibm.ws.security.openidconnect.server/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -168,6 +168,12 @@ requireRequestUriRegistration.desc=Indicate by true or false whether require req
+thirdPartyIDTokenClaims=Extra claims from a third-party ID Token
+thirdPartyIDTokenClaims.desc=The extra claims from a third-party ID Token to be put in the payload of the ID Token.
+
+thirdPartyAccessTokenClaims=Extra claims from a third-party Access Token
```
change
Access Token
to
access token

jantley-ibm commented on dev/com.ibm.ws.security.openidconnect.server/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -168,6 +168,12 @@ requireRequestUriRegistration.desc=Indicate by true or false whether require req
+thirdPartyIDTokenClaims.desc=The extra claims from a third-party ID Token to be put in the payload of the ID Token.
+
+thirdPartyAccessTokenClaims=Extra claims from a third-party Access Token
+thirdPartyAccessTokenClaims.desc=The extra claims from a third-party Access Token to be put in the payload of the Access Token.
```
change
Access Token
to
access token
(2 places)

jantley-ibm reviewed (commented):
Suggested edits from ID review attached, let me know if you have questions or comments. When complete I will add the "ID reviewed" tag

jimmy1wu commented:
#libby

jimmy1wu commented:
#build
#spawn.fullfat.buckets=com.ibm.ws.security.oidc.client_fat.1,com.ibm.ws.security.oidc.client_fat.2,com.ibm.ws.security.oidc.client_fat.config,com.ibm.ws.security.oidc.client_fat.e2e,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.saml,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat,com.ibm.ws.security.oidc.server_fat.app-passwords,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.clientregistration,com.ibm.ws.security.oidc.server_fat.config,com.ibm.ws.security.oidc.server_fat.config.oauth,com.ibm.ws.security.oidc.server_fat.config.oidc,com.ibm.ws.security.oidc.server_fat.endpoint.authorize,com.ibm.ws.security.oidc.server_fat.endpoint.coveragemap,com.ibm.ws.security.oidc.server_fat.endpoint.discovery,com.ibm.ws.security.oidc.server_fat.endpoint.end_session,com.ibm.ws.security.oidc.server_fat.endpoint.introspect,com.ibm.ws.security.oidc.server_fat.endpoint.revoke,com.ibm.ws.security.oidc.server_fat.endpoint.token,com.ibm.ws.security.oidc.server_fat.endpoint.userinfo,com.ibm.ws.security.oidc.server_fat.feature,com.ibm.ws.security.oidc.server_fat.granttype.jwt,com.ibm.ws.security.oidc.server_fat.gui.tokenMgmt,com.ibm.ws.security.oidc.server_fat.jaxrs,com.ibm.ws.security.oidc.server_fat.jaxrs.config,com.ibm.ws.security.oidc.server_fat.oauth,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.oidc.server_fat.saml,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP.1,com.ibm.ws.security.social_fat.LibertyOP.2,com.ibm.ws.security.social_fat.OpenShift,com.ibm.ws.security.social_fat.delegated,com.ibm.ws.security.social_fat.multiProvider

ayoho commented on dev/com.ibm.ws.security.oauth/src/com/ibm/oauth/core/internal/oauth20/granttype/impl/OAuth20GrantTypeHandlerRefreshImpl.java:
```diff
@@ -270,8 +271,13 @@ long getNumberOfRefreshedAccessTokens(Collection<OAuth20Token> userAndClientToke
+                    SubjectHelper subjectHelper = new SubjectHelper();
+                    String thirdPartyAccessToken = subjectHelper.getAccessTokenFromRunAsSubject();
+                    if (thirdPartyAccessToken != null) {
+                        tokenMap.put(OAuth20Constants.THIRD_PARTY_ACCESS_TOKEN_SUFFIX, new String[] { thirdPartyAccessToken });
```
Is `OAuth20Constants.THIRD_PARTY_ACCESS_TOKEN` supposed to be the key here?

jimmy1wu replied:
yes, thanks for catching that

ayoho commented on dev/com.ibm.ws.security.oauth/src/com/ibm/oauth/core/internal/oauth20/responsetype/impl/OAuth20ResponseTypeHandlerCodeImpl.java:
```diff
@@ -112,6 +116,55 @@ public void validateRequestResponseType(AttributeList attributeList, JsonArray r
             result = new ArrayList<OAuth20Token>();
             result.add(code);
 
+            SubjectHelper subjectHelper = new SubjectHelper();
```
If we have a chance to come back to this, this new section would probably be better suited in a separate method if at all possible to try to keep this method from growing too big.

ayoho commented on dev/com.ibm.ws.security.oauth/src/com/ibm/ws/security/oauth20/plugins/BaseTokenHandler.java:
```diff
@@ -158,6 +166,29 @@ public OAuth20Token createToken(Map<String, String[]> tokenMap) {
+                }
+            }
+        } catch (Exception e) {
+            // ignore for now
```
We'll probably want to consider adding an NLS warning message in here.

jimmy1wu replied:
an nls warning will be added in a follow-up pr

ayoho commented on dev/com.ibm.ws.security.openidconnect.server/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -34,6 +34,10 @@
             required="false" type="Boolean" default="true"/>
          <AD id="customClaims" name="%customClaims" description="%customClaims.desc"
             required="false" type="String" cardinality="2147483647"/>            
+         <AD id="thirdPartyIDTokenClaims" name="thirdPartyIDTokenClaims" description="thirdPartyIDTokenClaims.desc"
+            required="false" type="String" cardinality="2147483647"/>
+         <AD id="thirdPartyAccessTokenClaims" name="thirdPartyAccessTokenClaims" description="thirdPartyAccessTokenClaims.desc"
+            required="false" type="String" cardinality="2147483647"/>
```
Both of these need to be marked `ibm:beta="true"` until we have testing in place for them.

ayoho commented on dev/com.ibm.ws.security.openidconnect.server/src/com/ibm/ws/security/openidconnect/server/plugins/IDTokenHandler.java:
```diff
@@ -356,34 +360,69 @@ private void addOptionalClaims(Payload payload, @Sensitive Map<String, String[]>
+                }
+            }
+        } catch (Exception e) {
+            // ignore for now
```
Similar comment as above - should probably consider adding an NLS message here.

jimmy1wu replied:
ditto

ayoho commented on dev/com.ibm.ws.security.openidconnect.server/src/com/ibm/ws/security/openidconnect/server/plugins/IDTokenHandler.java:
```diff
@@ -356,34 +360,69 @@ private void addOptionalClaims(Payload payload, @Sensitive Map<String, String[]>
         return new HashMap<String, Object>();
     }
 
+    private Map<String, Object> getThirdPartyIDTokenClaims(String thirdPartyIDToken, OidcServerConfig oidcServerConfig) {
```
This is similar enough to the `getThirdPartyClaims()` method you added in `BaseTokenHandler` that you might be able to make a common method that they could share. Perhaps something like:
```Java
private Map<String, Object> getThirdPartyClaims(String thirdPartyToken, Set<String> allowedThirdPartyClaims) {
    Map<String, Object> thirdPartyTokenClaims = new HashMap<String, Object>();
    if (thirdPartyToken == null || thirdPartyToken.isEmpty()) {
        return thirdPartyTokenClaims;
    }
    try {
        JwtContext jwtContext = Jose4jUtil.parseJwtWithoutValidation(thirdPartyToken);
        JwtClaims jwtClaims = jwtContext.getJwtClaims();
        Map<String, Object> jwtClaimsMap = jwtClaims.getClaimsMap();

        for (String thirdPartyClaim : allowedThirdPartyClaims) {
            if (jwtClaimsMap.containsKey(thirdPartyClaim)) {
                thirdPartyTokenClaims.put(thirdPartyClaim, jwtClaimsMap.get(thirdPartyClaim));
            }
        }
    } catch (Exception e) {
        // TODO - NLS message
    }
    return thirdPartyTokenClaims;
}
```

ayoho commented on dev/com.ibm.ws.security.openidconnect.clients.common/src/com/ibm/ws/security/openidconnect/client/jose4j/util/Jose4jUtil.java:
```diff
@@ -429,4 +431,25 @@ ProviderAuthenticationResult checkForReusedJwt(ConvergedClientConfig clientConfi
         return null;
     }
 
+    public static Map<String, Object> getThirdPartyClaims(String thirdPartyToken, Set<String> allowedThirdPartyClaims) {
```
I don't think this is ultimately the right class for this method. We can find a more appropriate class and move this method later.

ayoho commented on dev/com.ibm.ws.security.openidconnect.clients.common/test/com/ibm/ws/security/openidconnect/client/jose4j/util/Jose4jUtilTest.java:
```diff
@@ -421,6 +425,40 @@ public void testGetVerifyKey_getKeyFromJwkUrl() throws Exception {
+        assertTrue("Expected third-party claims map to contain claim 'test1'.", thirdPartyClaims.containsKey("test1"));
+        assertFalse("Expected third-party claims map to not contain claim 'test4'.", thirdPartyClaims.containsKey("test4"));
+    }
+
```
Another test that I think would be good to add that I believe Chris had mentioned earlier would be to have a claim in the allowed third party token claims list that _isn't_ in the token.

jimmy1wu commented:
#libby

jimmy1wu commented:
#build
#spawn.fullfat.buckets=com.ibm.ws.security.oidc.client_fat.1,com.ibm.ws.security.oidc.client_fat.2,com.ibm.ws.security.oidc.client_fat.config,com.ibm.ws.security.oidc.client_fat.e2e,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.saml,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat,com.ibm.ws.security.oidc.server_fat.app-passwords,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.clientregistration,com.ibm.ws.security.oidc.server_fat.config,com.ibm.ws.security.oidc.server_fat.config.oauth,com.ibm.ws.security.oidc.server_fat.config.oidc,com.ibm.ws.security.oidc.server_fat.endpoint.authorize,com.ibm.ws.security.oidc.server_fat.endpoint.coveragemap,com.ibm.ws.security.oidc.server_fat.endpoint.discovery,com.ibm.ws.security.oidc.server_fat.endpoint.end_session,com.ibm.ws.security.oidc.server_fat.endpoint.introspect,com.ibm.ws.security.oidc.server_fat.endpoint.revoke,com.ibm.ws.security.oidc.server_fat.endpoint.token,com.ibm.ws.security.oidc.server_fat.endpoint.userinfo,com.ibm.ws.security.oidc.server_fat.feature,com.ibm.ws.security.oidc.server_fat.granttype.jwt,com.ibm.ws.security.oidc.server_fat.gui.tokenMgmt,com.ibm.ws.security.oidc.server_fat.jaxrs,com.ibm.ws.security.oidc.server_fat.jaxrs.config,com.ibm.ws.security.oidc.server_fat.oauth,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.oidc.server_fat.saml,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP.1,com.ibm.ws.security.social_fat.LibertyOP.2,com.ibm.ws.security.social_fat.OpenShift,com.ibm.ws.security.social_fat.delegated,com.ibm.ws.security.social_fat.multiProvider

jimmy1wu commented:
#libby

jimmy1wu commented:
#build
#spawn.fullfat.buckets=com.ibm.ws.security.oidc.client_fat.1,com.ibm.ws.security.oidc.client_fat.2,com.ibm.ws.security.oidc.client_fat.config,com.ibm.ws.security.oidc.client_fat.e2e,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.saml,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat,com.ibm.ws.security.oidc.server_fat.app-passwords,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.clientregistration,com.ibm.ws.security.oidc.server_fat.config,com.ibm.ws.security.oidc.server_fat.config.oauth,com.ibm.ws.security.oidc.server_fat.config.oidc,com.ibm.ws.security.oidc.server_fat.endpoint.authorize,com.ibm.ws.security.oidc.server_fat.endpoint.coveragemap,com.ibm.ws.security.oidc.server_fat.endpoint.discovery,com.ibm.ws.security.oidc.server_fat.endpoint.end_session,com.ibm.ws.security.oidc.server_fat.endpoint.introspect,com.ibm.ws.security.oidc.server_fat.endpoint.revoke,com.ibm.ws.security.oidc.server_fat.endpoint.token,com.ibm.ws.security.oidc.server_fat.endpoint.userinfo,com.ibm.ws.security.oidc.server_fat.feature,com.ibm.ws.security.oidc.server_fat.granttype.jwt,com.ibm.ws.security.oidc.server_fat.gui.tokenMgmt,com.ibm.ws.security.oidc.server_fat.jaxrs,com.ibm.ws.security.oidc.server_fat.jaxrs.config,com.ibm.ws.security.oidc.server_fat.oauth,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.oidc.server_fat.saml,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP.1,com.ibm.ws.security.social_fat.LibertyOP.2,com.ibm.ws.security.social_fat.OpenShift,com.ibm.ws.security.social_fat.delegated,com.ibm.ws.security.social_fat.multiProvider

skasund commented:
L2 message review not required

jantley-ibm commented on dev/com.ibm.ws.security.openidconnect.server/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -166,7 +166,13 @@ requireRequestUriRegistration=Require request URI registration
 
 customClaims=Extra claims of ID token
-customClaims.desc=The extra claims to be put in the payloads of the ID Token, in addition to the default realmName, uniqueSecurityName and groupIds claims.
+customClaims.desc=The extra claims to be put in the payloads of the ID token, in addition to the default realmName, uniqueSecurityName and groupIds claims.
```
Add a comma after uniqueSecurityName, like so:
to the default realmName, uniqueSecurityName, and groupIds claims.

jimmy1wu commented:
#build
#spawn.fullfat.buckets=com.ibm.ws.security.oidc.client_fat.1,com.ibm.ws.security.oidc.client_fat.2,com.ibm.ws.security.oidc.client_fat.config,com.ibm.ws.security.oidc.client_fat.e2e,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.saml,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat,com.ibm.ws.security.oidc.server_fat.app-passwords,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.clientregistration,com.ibm.ws.security.oidc.server_fat.config,com.ibm.ws.security.oidc.server_fat.config.oauth,com.ibm.ws.security.oidc.server_fat.config.oidc,com.ibm.ws.security.oidc.server_fat.endpoint.authorize,com.ibm.ws.security.oidc.server_fat.endpoint.coveragemap,com.ibm.ws.security.oidc.server_fat.endpoint.discovery,com.ibm.ws.security.oidc.server_fat.endpoint.end_session,com.ibm.ws.security.oidc.server_fat.endpoint.introspect,com.ibm.ws.security.oidc.server_fat.endpoint.revoke,com.ibm.ws.security.oidc.server_fat.endpoint.token,com.ibm.ws.security.oidc.server_fat.endpoint.userinfo,com.ibm.ws.security.oidc.server_fat.feature,com.ibm.ws.security.oidc.server_fat.granttype.jwt,com.ibm.ws.security.oidc.server_fat.gui.tokenMgmt,com.ibm.ws.security.oidc.server_fat.jaxrs,com.ibm.ws.security.oidc.server_fat.jaxrs.config,com.ibm.ws.security.oidc.server_fat.oauth,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.oidc.server_fat.saml,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP.1,com.ibm.ws.security.social_fat.LibertyOP.2,com.ibm.ws.security.social_fat.OpenShift,com.ibm.ws.security.social_fat.delegated,com.ibm.ws.security.social_fat.multiProvider

jimmy1wu committed:
address feedback

jimmy1wu commented:
scope of feature reduced to just id tokens. thus, i updated the pr to remove the commits related to propagating third party access token claims combined the rest. in terms of id, i removed the metatype.properties text related to the third party access token claims field i introduced earlier in this pr. the rest of the metatype changes have already been reviewed, so id doesn't have too re-review this pr.

jimmy1wu commented:
#libby

jimmy1wu commented:
#build
#spawn.fullfat.buckets=com.ibm.ws.security.oidc.client_fat.1,com.ibm.ws.security.oidc.client_fat.2,com.ibm.ws.security.oidc.client_fat.config,com.ibm.ws.security.oidc.client_fat.e2e,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.saml,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat,com.ibm.ws.security.oidc.server_fat.app-passwords,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.clientregistration,com.ibm.ws.security.oidc.server_fat.config,com.ibm.ws.security.oidc.server_fat.config.oauth,com.ibm.ws.security.oidc.server_fat.config.oidc,com.ibm.ws.security.oidc.server_fat.endpoint.authorize,com.ibm.ws.security.oidc.server_fat.endpoint.coveragemap,com.ibm.ws.security.oidc.server_fat.endpoint.discovery,com.ibm.ws.security.oidc.server_fat.endpoint.end_session,com.ibm.ws.security.oidc.server_fat.endpoint.introspect,com.ibm.ws.security.oidc.server_fat.endpoint.revoke,com.ibm.ws.security.oidc.server_fat.endpoint.token,com.ibm.ws.security.oidc.server_fat.endpoint.userinfo,com.ibm.ws.security.oidc.server_fat.feature,com.ibm.ws.security.oidc.server_fat.granttype.jwt,com.ibm.ws.security.oidc.server_fat.gui.tokenMgmt,com.ibm.ws.security.oidc.server_fat.jaxrs,com.ibm.ws.security.oidc.server_fat.jaxrs.config,com.ibm.ws.security.oidc.server_fat.oauth,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.oidc.server_fat.saml,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP.1,com.ibm.ws.security.social_fat.LibertyOP.2,com.ibm.ws.security.social_fat.OpenShift,com.ibm.ws.security.social_fat.delegated,com.ibm.ws.security.social_fat.multiProvider

jimmy1wu commented:
#build
#spawn.fullfat.buckets=com.ibm.ws.security.oidc.client_fat.1,com.ibm.ws.security.oidc.client_fat.2,com.ibm.ws.security.oidc.client_fat.config,com.ibm.ws.security.oidc.client_fat.e2e,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.saml,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat,com.ibm.ws.security.oidc.server_fat.app-passwords,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.clientregistration,com.ibm.ws.security.oidc.server_fat.config,com.ibm.ws.security.oidc.server_fat.config.oauth,com.ibm.ws.security.oidc.server_fat.config.oidc,com.ibm.ws.security.oidc.server_fat.endpoint.authorize,com.ibm.ws.security.oidc.server_fat.endpoint.coveragemap,com.ibm.ws.security.oidc.server_fat.endpoint.discovery,com.ibm.ws.security.oidc.server_fat.endpoint.end_session,com.ibm.ws.security.oidc.server_fat.endpoint.introspect,com.ibm.ws.security.oidc.server_fat.endpoint.revoke,com.ibm.ws.security.oidc.server_fat.endpoint.token,com.ibm.ws.security.oidc.server_fat.endpoint.userinfo,com.ibm.ws.security.oidc.server_fat.feature,com.ibm.ws.security.oidc.server_fat.granttype.jwt,com.ibm.ws.security.oidc.server_fat.gui.tokenMgmt,com.ibm.ws.security.oidc.server_fat.jaxrs,com.ibm.ws.security.oidc.server_fat.jaxrs.config,com.ibm.ws.security.oidc.server_fat.oauth,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.oidc.server_fat.saml,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP.1,com.ibm.ws.security.social_fat.LibertyOP.2,com.ibm.ws.security.social_fat.OpenShift,com.ibm.ws.security.social_fat.delegated,com.ibm.ws.security.social_fat.multiProvider

jimmy1wu committed:
fix npe and typo

jimmy1wu commented:
#build
#spawn.fullfat.buckets=com.ibm.ws.security.oidc.client_fat.1,com.ibm.ws.security.oidc.client_fat.2,com.ibm.ws.security.oidc.client_fat.config,com.ibm.ws.security.oidc.client_fat.e2e,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.saml,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat,com.ibm.ws.security.oidc.server_fat.app-passwords,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.clientregistration,com.ibm.ws.security.oidc.server_fat.config,com.ibm.ws.security.oidc.server_fat.config.oauth,com.ibm.ws.security.oidc.server_fat.config.oidc,com.ibm.ws.security.oidc.server_fat.endpoint.authorize,com.ibm.ws.security.oidc.server_fat.endpoint.coveragemap,com.ibm.ws.security.oidc.server_fat.endpoint.discovery,com.ibm.ws.security.oidc.server_fat.endpoint.end_session,com.ibm.ws.security.oidc.server_fat.endpoint.introspect,com.ibm.ws.security.oidc.server_fat.endpoint.revoke,com.ibm.ws.security.oidc.server_fat.endpoint.token,com.ibm.ws.security.oidc.server_fat.endpoint.userinfo,com.ibm.ws.security.oidc.server_fat.feature,com.ibm.ws.security.oidc.server_fat.granttype.jwt,com.ibm.ws.security.oidc.server_fat.gui.tokenMgmt,com.ibm.ws.security.oidc.server_fat.jaxrs,com.ibm.ws.security.oidc.server_fat.jaxrs.config,com.ibm.ws.security.oidc.server_fat.oauth,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.oidc.server_fat.saml,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP.1,com.ibm.ws.security.social_fat.LibertyOP.2,com.ibm.ws.security.social_fat.OpenShift,com.ibm.ws.security.social_fat.delegated,com.ibm.ws.security.social_fat.multiProvider

jimmy1wu commented:
#libby
